trunk/SUAVE/Components/Energy/Converters/Rotor.py

```diff
@@ -140,152 +150,220 @@ def spin(self,conditions):
         V[V==0.0] = 1E-6
-        ut        = np.zeros_like(V) 
-    
+        
         #Things that don't change with iteration
         Nr       = len(c) # Number of stations radially    
         ctrl_pts = len(Vv)
                  
-        # set up non dimensional radial distribution  
-        chi            = self.radius_distribution/R 
-        
-        omega          = np.abs(omega)        
-        r              = chi*R          # Radial coordinate 
-        pi             = np.pi        
-        pi2            = pi*pi       
-        n              = omega/(2.*pi)  # Cycles per second  
-        nu             = mu/rho         
+        # Non-dimensional radial distribution and differential radius
+        chi           = r_1d/R 
+        diff_r        = np.diff(r_1d)
+        deltar        = np.zeros(len(r_1d))
+        deltar[1:-1]  = diff_r[0:-1]/2 + diff_r[1:]/2
+        deltar[0]     = diff_r[0]/2
+        deltar[-1]    = diff_r[-1]/2
+        
+        # Check for negative rotation rate
+        if np.any(omega) <0:
+            print("Negative rotation rate defined! Reverting direction.")
+            omega    = np.abs(omega)
+            rotation = -rotation        
+        
+        # Calculating rotational parameters
+        omegar   = np.outer(omega,r_1d)    
+        n        = omega/(2.*pi)   # Rotations per second  
+             
     
-        # azimuth distribution 
+        # Azimuthal distribution of stations
         psi            = np.linspace(0,2*pi,Na+1)[:-1]
         psi_2d         = np.tile(np.atleast_2d(psi).T,(1,Nr))
         psi_2d         = np.repeat(psi_2d[np.newaxis, :, :], ctrl_pts, axis=0)   
-        azimuth_2d     = np.repeat(np.atleast_2d(psi).T[np.newaxis,: ,:], Na, axis=0).T  
         
         # 2 dimensiona radial distribution non dimensionalized
         chi_2d         = np.tile(chi ,(Na,1))            
-        chi_2d         = np.repeat(chi_2d[ np.newaxis,:, :], ctrl_pts, axis=0) 
-                       
-        r_dim_2d       = np.tile(r ,(Na,1))  
-        r_dim_2d       = np.repeat(r_dim_2d[ np.newaxis,:, :], ctrl_pts, axis=0)  
+        chi_2d         = np.repeat(chi_2d[None,:,:], ctrl_pts, axis=0) 
+        r_dim_2d       = np.tile(r_1d ,(Na,1))  
+        r_dim_2d       = np.repeat(r_dim_2d[None,:,:], ctrl_pts, axis=0)          
+        c_2d           = np.tile(c ,(Na,1)) 
+        c_2d           = np.repeat(c_2d[None,:,:], ctrl_pts, axis=0)  
+        
+        # Starting with uniform freestream
+        ua       = 0             
+        ut       = 0            
+        ur       = 0             
     
-        # Things that will change with iteration
-        size   = (len(a),Nr)
-        omegar = np.outer(omega,r)
-        Ua     = np.outer((V),np.ones_like(r))
+        # Include velocities introduced by rotor incidence angles
+        if not np.all(np.array(self.orientation_euler_angles)==0): # this needs to be instead angle to freestream, not to body
```

Comment:
I was hoping you'd update this

Reply:
yes it's on my list. I am working on adding this and also changing a lot of other files to account for the identical_propellers capabilities